File: rb120_oop/lesson_5/twenty_one.rb
module Hand
  MAX_VALUE = 21
  def show_hand
    puts "#{name}'s Hand:"
    cards.each do |card|
      puts "=> #{card}"
    end
    puts "=> Total: #{total}"
    puts ""
  end

  def total
    total = 0
    cards.each do |card|
      total += if card.ace?
                 11
               elsif card.jack? || card.queen? || card.king?
                 10
               else
                 card.value.to_i
               end
    end

    cards.select(&:ace?).count.times do
      break if total <= MAX_VALUE
      total -= 10
    end
    total
  end

  def add_card(card)
    cards << card
  end

  def busted?
    total > MAX_VALUE
  end
end

class Participant
  include Hand
  attr_accessor :name, :cards, :score

  def initialize
    @cards = []
    @score = 0
    set_name
  end
end

class Player < Participant
  def show_flop
    show_hand
  end

  private

  def set_name
    name = ''
    loop do
      puts "What's your name?"
      name = gets.chomp.strip
      break unless name.size < 3
      puts "Sorry, name must be at least three characters."
    end
    self.name = name
  end
end

class Dealer < Participant
  def show_flop
    puts "#{name}'s Hand:"
    puts "=> #{cards.first}"
    puts "=> ?? "
    puts ""
  end

  private

  def set_name
    self.name = 'DeepMind'
  end
end

class Deck
  attr_accessor :cards
  def initialize
    @cards = []
    Card::SUITS.each do |suit|
      Card::VALUES.each do |value|
        @cards << Card.new(suit, value)
      end
    end
    @cards.shuffle!
  end

  def deal_one
    cards.pop
  end
end

class Card
  SUITS = %w(D S H C)
  VALUES = %w(2 3 4 5 6 7 8 9 10 J Q K A)

  def initialize(suit, value)
    @suit = suit
    @value = value
  end

  def to_s
    "#{value} of #{suit}"
  end

  def suit
    case @suit
    when 'D' then 'Diamonds'
    when 'S' then 'Spades'
    when 'H' then 'Hearts'
    when 'C' then 'Clubs'
    end
  end

  def value
    case @value
    when 'J' then 'Jack'
    when 'Q' then 'Queen'
    when 'K' then 'King'
    when 'A' then 'Ace'
    else @value
    end
  end

  def ace?
    value == 'Ace'
  end

  def king?
    value == 'King'
  end

  def queen?
    value == 'Queen'
  end

  def jack?
    value == 'Jack'
  end
end

class Game
  DEALER_MIN = 17
  NUM_OF_WINS = 3
  attr_accessor :deck, :dealer, :player

  def initialize
    system 'clear'
    @deck = Deck.new
    @dealer = Dealer.new
    @player = Player.new
  end

  def start
    greeting
    loop do
      reset_score
      play_round
      break unless play_again?
    end
  end

  private

  def play_round
    loop do
      shuffle_and_deal

      player_turn
      if player.busted?
        calculate_and_display_result
        break if game_over?
        next
      end

      dealer_turn

      show_hands
      calculate_and_display_result
      break if game_over?
    end
  end

  def calculate_and_display_result
    determine_result
    update_score
    display_winner
  end

  def game_over?
    player.score >= NUM_OF_WINS || dealer.score >= NUM_OF_WINS
  end

  def shuffle_and_deal
    reset
    deal_cards
    show_flop
  end

  def continue
    puts "Press enter to continue:"
    gets
  end

  def greeting
    system 'clear'
    puts "Welcome to Twenty-One! This is a simplified version of the very
    popular card game called Blackjack. First to 3 wins. Have fun!"
    puts ''
    continue
  end

  def request_action
    answer = nil
    loop do
      puts "Would you like to (h)it or (s)tay?"
      answer = gets.chomp.downcase
      break if ['h', 'hit', 's', 'stay'].include?(answer)
      puts "Sorry, not a valid response."
    end
    answer
  end

  def stay?(answer)
    answer == "s" || answer == "stay"
  end

  def player_stays
    puts "#{player.name} stays!"
    continue
    system 'clear'
  end

  def player_hits
    puts "#{player.name} hits!"
    new_card = deck.deal_one
    player.add_card(new_card)
    sleep 1
    puts "#{player.name} draws a #{new_card.value} of #{new_card.suit}."
    sleep 1
    continue
    system 'clear'
  end

  def player_turn
    loop do
      answer = request_action
      system 'clear'
      if stay?(answer)
        player_stays
        break
      elsif player.busted?
        break
      else
        player_hits
        player.show_hand
        break if player.busted?
      end
    end
  end

  def dealer_turn
    loop do
      dealer.show_hand
      sleep 1
      break if dealer.total >= DEALER_MIN
      puts "#{dealer.name} draws a card."
      sleep 1
      dealer.add_card(deck.deal_one)
    end
    continue
    system 'clear'
  end

  def play_again?
    puts "Would you like to play again? (y or n)"
    answer = nil
    loop do
      answer = gets.chomp.downcase
      break if ['y', 'yes', 'no', 'n'].include?(answer)
      puts "Sorry, not a valid response."
    end
    answer == 'y' || answer == 'yes'
  end

  def reset
    system 'clear'
    self.deck = Deck.new
    player.cards = []
    dealer.cards = []
  end

  def deal_cards
    system 'clear'
    puts "Shuffling and dealing cards..."
    2.times do
      dealer.add_card(deck.deal_one)
      player.add_card(deck.deal_one)
    end
    sleep 3
  end

  def show_flop
    system 'clear'
    player.show_flop
    sleep 1
    dealer.show_flop
  end

  def show_hands
    system 'clear'
    player.show_hand
    sleep 1
    dealer.show_hand
    sleep 1
  end

  def determine_result
    dealer_name = dealer.name
    player_name = player.name
    dealer_total = dealer.total
    player_total = player.total
    output = if dealer.busted?
               "#{dealer_name} busted! #{player_name} wins!"
             elsif player.busted?
               "#{player_name} busted! #{dealer_name} wins!"
             elsif player_total > dealer_total
               "#{player_name} wins!"
             elsif dealer_total > player_total
               "#{dealer_name} wins!"
             else
               "It's a tie!"
             end
    puts output
  end

  def reset_score
    player.score = 0
    dealer.score = 0
  end

  def calibrate(total)
    total > Hand::MAX_VALUE ? (total % Hand::MAX_VALUE) : total
  end

  def update_score
    player_total = calibrate(player.total)
    dealer_total = calibrate(dealer.total)
    if dealer.busted? || (player_total > dealer_total)
      player.score += 1
    elsif player.busted? || (dealer_total > player_total)
      dealer.score += 1
    end
  end

  def display_score
    player_score = player.score
    dealer_score = dealer.score
    output = if player_score > dealer_score
               "#{player.name} is winning #{player_score}:#{dealer_score}!"
             elsif dealer_score > player_score
               "#{dealer.name} is winning #{dealer_score}:#{player_score}!"
             else
               "The score is tied #{player_score}:#{dealer_score}!"
             end
    puts output
    continue
  end

  def display_winner
    player_score = player.score
    dealer_score = dealer.score
    if player_score >= NUM_OF_WINS
      puts "#{player.name} wins #{player_score}:#{dealer_score}!"
    elsif dealer_score >= NUM_OF_WINS
      puts "#{dealer.name} wins #{dealer_score}:#{player_score}!"
    else
      display_score
    end
  end
end

Game.new.start

File: js101_foundations/lesson_6/twenty_one.js
const readline = require('readline-sync');
const MSG = require('./twenty_one.json');

const SUITS = ['s', 'c', 'h', 'd'];
const VALUES = ['A', '2', '3', '4', '5', '6', '7', '8', '9', '10', 'J', 'Q', 'K'];
const WINNING_COUNT = 3;
const BLACKJACK = 21;

function initializeDeck() {
  let deck = [];

  SUITS.forEach(suit => {
    VALUES.forEach(value => {
      deck.push([value, suit]);
    });
  });
  shuffle(deck);
  return deck;
}

function prompt(msg) {
  console.log(`=> ${msg}`);
}

function lineBreak() {
  console.log("------------------------------------");
}

function welcome() {
  console.clear();
  prompt(MSG["welcome"]);
  console.log(`${MSG["numberOfWins"]}`, WINNING_COUNT);
  askToContinue();
}

function shuffle(deck) {
  for (let idx = deck.length - 1; idx > 0; idx--) {
    let otherIdx = Math.floor(Math.random() * (idx + 1));
    [deck[idx], deck[otherIdx]] = [deck[otherIdx], deck[idx]];
  }
}

function getPlayerAction() {
  console.log(" ");
  prompt(MSG["hitOrStay"]);
  return readline.question().toLowerCase();
}

function getValidAction(answer) {
  while (isInvalidAction(answer)) {
    prompt(MSG["invalidAnswer"]);
    answer = readline.question().toLowerCase();
  }
  return answer;
}

function askToContinue() {
  lineBreak();
  prompt(MSG["continue"]);
  readline.question();
}

function dealCards(deck, player, dealer) {
  for (let count = 0; count < 2; count++) {
    player.push(deck.pop());
    dealer.push(deck.pop());
  }
}

function dealCard(deck, player) {
  console.clear();
  player.push(deck.pop());
}

function isBusted(total) {
  return total > BLACKJACK;
}

function isWon(firstTotal, secondTotal) {
  return firstTotal > secondTotal;
}

function total(cards) {
  let values = cards.map(card => card[0]);
  let sum = 0;

  values.forEach(value => {
    if (value === "A") {
      sum += 11;
    } else if (['J', 'Q', 'K'].includes(value)) {
      sum += 10;
    } else {
      sum += Number(value);
    }
  });

  values.filter(value => value === "A").forEach(_ => {
    if (isBusted(sum)) sum -= 10;
  });

  return sum;
}

function displayCards(player) {
  return player.map(card => card.join("")).join(" ");
}

function displayTotals(player, dealer, playerTurn = true) {
  console.clear();
  prompt(`Player: ${displayCards(player)}  |  Total: ${total(player)}`);
  lineBreak();
  if (playerTurn) {
    prompt(`Dealer: ${dealer[0].join('')} ?   |  Total: ?`);
  } else {
    prompt(`Dealer: ${displayCards(dealer)}  |  Total: ${total(dealer)}`);
    askToContinue();
  }
}

function displayDealtCard(player, playerTurn = true) {
  if (playerTurn) {
    prompt(MSG["playerHits"]);
    prompt(`You draw a ${player[player.length - 1].join("")}.`);
  } else {
    prompt(MSG["dealerHits"]);
    prompt(`Dealer draws a ${player[player.length - 1].join("")}.`);
  }
  askToContinue();
}

function isInvalidAnswer(input) {
  return !['y', 'yes', 'n', 'no'].includes(input);
}

function playerStays(answer) {
  return ['s', 'stay'].includes(answer);
}

function displayStay() {
  console.clear();
  prompt(MSG["playerStays"]);
  askToContinue();
}

function displayResults(playerTotal, dealerTotal) {
  console.clear();

  if (isBusted(playerTotal)) {
    prompt(MSG["playerBusts"]);
  } else if (isBusted(dealerTotal)) {
    prompt(MSG["dealerBusts"]);
  } else if (isWon(playerTotal, dealerTotal)) {
    prompt(`You win ${playerTotal}:${dealerTotal}!`);
  } else if (isWon(dealerTotal, playerTotal)) {
    prompt(`Dealer wins ${dealerTotal}:${playerTotal}!`);
  } else {
    prompt(MSG["tie"]);
  }
}

function playTurnAfterHit(deck, player, dealer, playerTurn = true) {
  if (playerTurn) {
    dealCard(deck, player);
    displayDealtCard(player);
    displayTotals(player, dealer);
  } else {
    dealCard(deck, dealer);
    displayDealtCard(dealer, false);
    displayTotals(player, dealer, false);
  }
}

function displayTurnResults(score, player, dealer) {
  let playerTotal = total(player);
  let dealerTotal = total(dealer);

  displayResults(playerTotal, dealerTotal);
  updateScore(score, playerTotal, dealerTotal);
  displayScore(score);
}

function playAgain() {
  console.clear();
  prompt(MSG["playAgain"]);
  return readline.question().toLowerCase();
}

function askPlayAgain(again) {
  while (isInvalidAnswer(again)) {
    prompt(MSG["invalidAnswer"]);
    again = readline.question().toLowerCase();
  }

  return again;
}

function isInvalidAction(answer) {
  return !['hit', 'h', 'stay', 's'].includes(answer);
}

function isNo(again) {
  return ['n', 'no'].includes(again);
}

function updateScore(score, playerTotal, dealerTotal) {
  if (isBusted(playerTotal)) {
    score.dealer += 1;
  } else if (isBusted(dealerTotal)) {
    score.player += 1;
  } else if (isWon(playerTotal, dealerTotal)) {
    score.player += 1;
  } else if (isWon(dealerTotal, playerTotal)) {
    score.dealer += 1;
  }
}

function displayScore(score) {
  let playerScore = score.player;
  let dealerScore = score.dealer;

  if (dealerScore === WINNING_COUNT) {
    prompt(`Dealer wins ${dealerScore}:${playerScore}.`);
  } else if (playerScore === WINNING_COUNT) {
    prompt(`Player wins ${playerScore}:${dealerScore}.`);
  } else if (isWon(dealerScore, playerScore)) {
    prompt(`Dealer is winning ${dealerScore}:${playerScore}.`);
  } else if (isWon(playerScore, dealerScore)) {
    prompt(`You are winning ${playerScore}:${dealerScore}.`);
  } else {
    prompt(`You are tied ${playerScore}:${dealerScore}.`);
  }
  askToContinue();
}

function gameOver(score) {
  return score.player === WINNING_COUNT || score.dealer === WINNING_COUNT;
}

function goodbye() {
  prompt(MSG["goodbye"]);
}

welcome();

while (true) {
  let score = {player: 0, dealer: 0 };

  while (!gameOver(score)) {
    let deck = initializeDeck();
    let player = [];
    let dealer = [];

    dealCards(deck, player, dealer);
    displayTotals(player, dealer);

    while (true) {
      let answer = getPlayerAction();
      answer = getValidAction(answer);

      if (playerStays(answer)) {
        displayStay();
        displayTotals(player, dealer, false);
        break;
      }

      playTurnAfterHit(deck, player, dealer);
      if (isBusted(total(player))) break;
    }

    if (isBusted(total(player))) {
      displayTotals(player, dealer, false);
      displayTurnResults(score, player, dealer);
    } else {
      while (total(dealer) < 17) {
        playTurnAfterHit(deck, player, dealer, false);
      }

    displayTurnResults(score, player, dealer);
    }
  }
  let again = playAgain();
  again = askPlayAgain(again);

  if (isNo(again)) break;
}

goodbye();